src/main/kotlin/com/bardsoftware/papeeria/backend/cosmas/PatchCorrector.kt

```diff
@@ -0,0 +1,69 @@
+/**
+Copyright 2018 BarD Software s.r.o
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+ */
+
+package com.bardsoftware.papeeria.backend.cosmas
+import java.util.LinkedList
+import name.fraser.neil.plaintext.diff_match_patch
+
+object PatchCorrector {
+    private fun getReverseOperation(operation: diff_match_patch.Operation): diff_match_patch.Operation {
+        return when (operation) {
+            diff_match_patch.Operation.DELETE -> diff_match_patch.Operation.INSERT
+            diff_match_patch.Operation.INSERT -> diff_match_patch.Operation.DELETE
+            else -> {
+                diff_match_patch.Operation.EQUAL
+            }
+        }
+    }
+
+    fun reversePatch(patchList: LinkedList<diff_match_patch.Patch>): LinkedList<diff_match_patch.Patch> {
+        val reversePatchList = LinkedList<diff_match_patch.Patch>()
+        for (patch in patchList) {
+            val newPatch = diff_match_patch.Patch()
+            newPatch.length1 = patch.length2
```

Comment:
Are you sure it is the right way of building the reverse patch? I am not. Here is a modified version of one of your tests where I added printing of patches:
```
    fun plusReversePatchTest() {
        val text1 = "Hello world!"
        val text2 = "Hello, world!"
        val dmp = diff_match_patch()
        val patch = dmp.patch_make(text1, text2)
      println("Original patch:\n $patch")
        val listOfPatch = LinkedList<diff_match_patch.Patch>()
        listOfPatch.addAll(PatchCorrector.reversePatch(patch))
      println("Home-grown reverse:\n $listOfPatch")
      println("Reverse built by DMP:\n ${dmp.patch_make(text2, text1)}")
        val newText = dmp.patch_apply(listOfPatch, text2)
        assertEquals(text1, newText[0])
    }
```

Here is the output:
```
Original patch:
 [@@ -1,12 +1,13 @@
 Hello
+,
  world!
]
Home-grown reverse:
 [@@ -1,13 +1,12 @@
 Hello
-,
  world!
]
Reverse built by DMP:
 [@@ -2,9 +2,8 @@
 ello
-,
  wor
]
```

You can see that home-grown reverse patch is significantly different from what DMP builds as patch which transitions text2 to text1.

If you are sure that such swaps of lengths and starts really produce a valdi reverse patch, please add a link to docs here (or write your own).

Reply:
ok, change